lib/perl/Genome/Model/Tools/Bsmap/methylation_ratio.py

```diff
@@ -0,0 +1,181 @@
+                ref_seq = genome.fetch(chromosome, start, end+3)
+            position = pileupcolumn.pos
+            if position < start or position >= end or pileupcolumn.n == 0:
+                continue
+            ref_seq_offset = position-start
+            if ref_seq[ref_seq_offset] in cg:
+                counter[position] = Counter()
+                counter[position].set_coverage(pileupcolumn.n)
+                counter[position].set_chromosome(chromosome)
+                meth_cov = 0
+                un_meth_cov = 0
+                for pileupread in pileupcolumn.pileups:          
+                    read = pileupread.alignment
+                    watson_strand = isOnWatsonStrand(read.tags)
+                    if read.query_sequence[pileupread.query_position] == ref_seq[ref_seq_offset]:
+                        if ref_seq[ref_seq_offset] == nuc_C and watson_strand:
+                            meth_cov += 1
+                        elif ref_seq[ref_seq_offset] == nuc_G and watson_strand is False:
+                            meth_cov += 1
+                    elif ref_seq[ref_seq_offset] == nuc_C and watson_strand and read.query_sequence[pileupread.query_position] == nuc_T:
+                        un_meth_cov += 1
+                    elif ref_seq[ref_seq_offset] == nuc_G and watson_strand is False and read.query_sequence[pileupread.query_position] == nuc_A:
+                        un_meth_cov += 1           
+                counter[position].set_methylation_coverage(meth_cov)
+                counter[position].set_un_methylation_coverage(un_meth_cov)
+        out_queue.put(counter)
+    print 'Quitting work'
+
+class Counter:
+    def __init__(self):
+        self.coverage = 0
+        self.methylation_coverage = 0
+        self.un_methylation_coverage = 0
+    def set_chromosome(self, chr):
+        self.chromosome = chr
+    def set_coverage(self, cov):
+        self.coverage = cov
+    def set_context(self, context):
+        self.context = context
+    def set_methylation_coverage(self, cov):
+        self.methylation_coverage = cov
+    def set_un_methylation_coverage(self, cov):
+        self.un_methylation_coverage = cov
+    def getCoverage(self):
+        return self.coverage
+    def getMethylationCoverage(self):
+        return self.methylation_coverage
+    def get_un_methylation_coverage(self):
+        return self.un_methylation_coverage
+    def get_chromosome(self):
+        return self.chromosome
+    def get_context(self):
+        return self.context
+
+if __name__ == '__main__':
+     
+    do_debug = False
+    usage = "usage: %prog [options] BSMAP_MAPPING_FILES (sorted BAM files)"
+    parser = optparse.OptionParser(usage=usage)
+    parser.add_option("-o", "--out", dest="outfile", metavar="FILE", help="output file name. (required)", default="")
+    parser.add_option("-d", "--ref", dest="reffile", metavar="FILE", help="reference genome fasta file. (required)", default="")
```

Comment:
I'm personally a fan of required options, I liken then to named arguments, but it's worth noting,

> Lots of people want their programs to have “required options”. Think about it. If it’s required, then it’s not optional! If there is a piece of information that your program absolutely requires in order to run successfully, that’s what positional arguments are for.

Source: [optparse](https://docs.python.org/2/library/optparse.html)

That also gets back to my comment earlier of the default output usually being stdout.